zebrad/src/commands/start.rs

```diff
@@ -67,3 +131,123 @@ impl config::Override<ZebradConfig> for StartCmd {
+            self.request_blocks(hashes).await?;
+
+            // Allow at most 300 block requests in flight.
+            self.drain_requests(300).await?;
+        }
+
+        self.drain_requests(0).await?;
+
+        let eternity = future::pending::<()>();
+        eternity.await;
+
+        Ok(())
+    }
+
+    async fn next_hashes(&mut self) -> Result<Vec<BlockHeaderHash>, Report> {
+        // Request the next 500 hashes.
+        self.retry_peer_set
+            .ready_and()
+            .await
+            .map_err(|e| eyre!(e))?
+            .call(zebra_network::Request::FindBlocks {
+                known_blocks: vec![self.tip],
+                stop: None,
+            })
+            .await
+            .map_err(|e| eyre!(e))
+            .wrap_err("request failed, TODO implement retry")
+            .map(|response| match response {
+                zebra_network::Response::BlockHeaderHashes(hashes) => hashes,
+                _ => unreachable!("FindBlocks always gets a BlockHeaderHashes response"),
+            })
+            .map(|hashes| {
+                info!(
+                    new_hashes = hashes.len(),
+                    requested = self.requested_block_heights,
+                    in_flight = self.block_requests.len(),
+                    downloaded = self.downloaded_block_heights.len(),
+                    highest = self.downloaded_block_heights.iter().next_back().unwrap().0,
+                    "requested more hashes"
+                );
+                self.requested_block_heights += hashes.len();
+                hashes
+            })
+    }
+
+    async fn request_blocks(&mut self, hashes: Vec<BlockHeaderHash>) -> Result<(), Report> {
+        for chunk in hashes.chunks(10usize) {
+            let request = self.peer_set.ready_and().await.map_err(|e| eyre!(e))?.call(
+                zebra_network::Request::BlocksByHash(chunk.iter().cloned().collect()),
+            );
+
+            self.block_requests.push(request);
+        }
+
+        Ok(())
+    }
+
+    async fn drain_requests(&mut self, request_goal: usize) -> Result<(), Report> {
+        while self.block_requests.len() > request_goal {
+            match self.block_requests.next().await {
+                Some(Ok(zebra_network::Response::Blocks(blocks))) => {
+                    for block in blocks {
+                        self.downloaded_block_heights
+                            .insert(block.coinbase_height().unwrap());
+                        self.state
+                            .ready_and()
+                            .await
+                            .map_err(|e| eyre!(e))?
+                            .call(zebra_state::Request::AddBlock { block })
+                            .await
+                            .map_err(|e| eyre!(e))?;
+                    }
+                }
+                Some(Err(e)) => {
+                    error!(%e);
+                }
+                _ => continue,
```

Comment:
`None` can never happen here, right, since `self.block_requests.len() > 0`?

Reply:
I think so, do you want me to change this to an `unreachable!`?